### PR TITLE
feat: expose `shouldUseFlatConfig`

### PR DESCRIPTION
**Pull request by cprussin**
#### Prerequisites checklist

- [x] I have read the [contributing guidelines](https://github.com/eslint/eslint/blob/HEAD/CONTRIBUTING.md).

#### What is the purpose of this pull request? (put an "X" next to an item)

[ ] Documentation update
[ ] Bug fix ([template](https://raw.githubusercontent.com/eslint/eslint/HEAD/templates/bug-report.md))
[ ] New rule ([template](https://raw.githubusercontent.com/eslint/eslint/HEAD/templates/rule-proposal.md))
[ ] Changes an existing rule ([template](https://raw.githubusercontent.com/eslint/eslint/HEAD/templates/rule-change-proposal.md))
[ ] Add autofix to a rule
[ ] Add a CLI option
[x] Add something to the core
[ ] Other, please explain:

#### What changes did you make? (Give an overview)

- Moved `shouldUseFlatConfig` from `lib/cli.js` to `lib/eslint/flat-eslint.js` and exposed it via exports
- Removed `findFlatConfigFile` from exports in `lib/eslint/flat-eslint.js` since it was only used to build `shouldUseFlatConfig` in `lib/cli.js`
- Exposed `shouldUseFlatConfig` via `lib/unsupported-apis.js`
- Added tests for `shouldUseFlatConfig`

#### Is there anything you'd like reviewers to focus on?

Nothing in particular beyond the typical code review

### CODE REVIEW

File: lib/eslint/flat-eslint.js [hunk 1]
```diff
@@ -1212,11 +1212,36 @@ class FlatESLint {
     }
 }
 
+/**
+ * Returns whether flat config should be used.
+ * @param {boolean} [allowFlatConfig] Whether or not to allow flat config.
+ * @returns {Promise<boolean>} Where flat config should be used.
+ */
+async function shouldUseFlatConfig(allowFlatConfig) {
```
Comment (nzakas) [inline lib/eslint/flat-eslint.js:1220]: I'm not sure it makes sense to expose the function with this argument. It was a convenience internally, but I think for external consumption, there would be no need to call `shouldUseFlatConfig()` if `allowFlatConfig` was `false`.

So, I think we should just get rid of the argument here and assume anyone calling `shouldUseFlatConfig()` wants to allow it.

**Review by nzakas — CHANGES_REQUESTED** [review]
Nice work. Just left a suggestion that we may want to change the function signature before exposing this publicly.

File: lib/eslint/flat-eslint.js (hunk 1 above)
Comment (cprussin) [reply lib/eslint/flat-eslint.js:1220]: yeah, that's fair enough, good catch and thanks!

cleans up the tests a bit too :)

Comment (cprussin) [reply lib/eslint/flat-eslint.js]: fixed now

File: lib/eslint/flat-eslint.js [hunk 2]
```diff
@@ -1212,11 +1212,31 @@ class FlatESLint {
     }
 }
 
+/**
+ * Returns whether flat config should be used.
+ * @returns {Promise<boolean>} Where flat config should be used.
```
Comment (nzakas) [inline lib/eslint/flat-eslint.js:1217]: Not your fault, but might as well fix it now.

```suggestion
 * @returns {Promise<boolean>} Whether flat config should be used.
```

**Review by nzakas — APPROVED** [review]
LGTM. Just one suggestion to clean up the comments.

Would like @mdjermanovic to review before merging.

File: lib/eslint/flat-eslint.js (hunk 2 above)
Comment (cprussin) [reply lib/eslint/flat-eslint.js]: Ah I hadn't even noticed that :)

done

**Review by nzakas — APPROVED** [review]

File: tests/lib/eslint/flat-eslint.js [hunk 3]
```diff
@@ -5448,3 +5449,67 @@ describe("FlatESLint", () => {
...
+        beforeEach(() => {
+            process.env.ESLINT_USE_FLAT_CONFIG = true;
+        });
+
+        afterEach(() => {
+            delete process.env.ESLINT_USE_FLAT_CONFIG;
+        });
+
+        it("is `true`", async () => {
+            assert.strictEqual(await shouldUseFlatConfig(true), true);
```
Comment (mdjermanovic) [inline tests/lib/eslint/flat-eslint.js:5464]:
```suggestion
            assert.strictEqual(await shouldUseFlatConfig(), true);
```

File: tests/lib/eslint/flat-eslint.js [hunk 4]
```diff
@@ -5448,3 +5449,67 @@ describe("FlatESLint", () => {
...
+        beforeEach(() => {
+            process.env.ESLINT_USE_FLAT_CONFIG = false;
+        });
+
+        afterEach(() => {
+            delete process.env.ESLINT_USE_FLAT_CONFIG;
+        });
+
+        it("is `false`", async () => {
+            assert.strictEqual(await shouldUseFlatConfig(true), false);
```
Comment (mdjermanovic) [inline tests/lib/eslint/flat-eslint.js:5478]:
```suggestion
            assert.strictEqual(await shouldUseFlatConfig(), false);
```

File: tests/lib/eslint/flat-eslint.js [hunk 5]
```diff
@@ -5448,3 +5449,67 @@ describe("FlatESLint", () => {
...
+            beforeEach(() => {
+                process.chdir(__dirname);
+            });
+
+            afterEach(() => {
+                process.chdir(originalDir);
+            });
+
+            it("is `true`", async () => {
+                assert.strictEqual(await shouldUseFlatConfig(true), true);
```
Comment (mdjermanovic) [inline tests/lib/eslint/flat-eslint.js:5495]:
```suggestion
                assert.strictEqual(await shouldUseFlatConfig(), true);
```

File: tests/lib/eslint/flat-eslint.js [hunk 6]
```diff
@@ -5448,3 +5449,67 @@ describe("FlatESLint", () => {
...
+            beforeEach(() => {
+                process.chdir(os.tmpdir());
+            });
+
+            afterEach(() => {
+                process.chdir(originalDir);
+            });
+
+            it("is `false`", async () => {
+                assert.strictEqual(await shouldUseFlatConfig(true), false);
```
Comment (mdjermanovic) [inline tests/lib/eslint/flat-eslint.js:5511]:
```suggestion
                assert.strictEqual(await shouldUseFlatConfig(), false);
```

File: tests/lib/eslint/flat-eslint.js [hunk 7]
```diff
@@ -5448,3 +5449,67 @@ describe("FlatESLint", () => {
     });
 
 });
+
+describe("shouldUseFlatConfig", () => {
+    describe("when the env variable `ESLINT_USE_FLAT_CONFIG` is `'true'`", () => {
+        beforeEach(() => {
+            process.env.ESLINT_USE_FLAT_CONFIG = true;
+        });
```
Comment (mdjermanovic) [inline tests/lib/eslint/flat-eslint.js:5457]: I think this test would be better with a setup where the is no config file present (e.g., with `process.chdir(os.tmpdir())`).

Comment (cprussin) [reply tests/lib/eslint/flat-eslint.js]: Great suggestion, I refactored the test to now test the entire cross product space of env var values x directory, so we should be covered now

**Comment by cprussin** [comment]
thanks for the review @mdjermanovic , I've addressed all your comments -- let me know if there's anything else I can clean up!

**Review by mdjermanovic — APPROVED** [review]
LGTM, thanks for contributing!

**Pull request merged**